File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n=nums.length;
        int[] prefix = new int[n];
        Arrays.fill(prefix,1);
        int count=1;
        for(int i=0; i<n; i++){
            prefix[i] *= count;
            count *= nums[i];
        }
        int[] ans = new int[n];
        Arrays.fill(ans, 1);
        count=1;
        for(int i=n-1; i>=0; i--){
            ans[i] = prefix[i]*count;
            count *= nums[i];
        }
        return ans;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        prod=[1]*n
        prev=nums[0]
        for i, num in enumerate(nums):
            if i==0:
                continue
            prod[i] = prod[i-1]*prev
            prev=num
        suff=1
        for i in range(n-1, -1, -1):
            if i==n-1:
                continue
            suff *= nums[i+1]
            prod[i] *= suff
        return prod